org.eclipse.scanning.api/src/org/eclipse/scanning/api/database/IExperimentDatabaseService.java

```diff
@@ -1,160 +1,160 @@
+ *    See git history
+ *******************************************************************************/
+package org.eclipse.scanning.api.database;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.concurrent.Future;
+
+/**
+ * 
+ * This is a service to bridge between the needs of 
+ * experimental information and the features of the API's.
+ * 
+ * The database we refer to here is 'Ispyb' although any
+ * database recording experimental data could implement this
+ * front end.
+ * 
+ * This service is designed to be exposed as a microservice
+ * operating over activemq. It may either be incorporated directly
+ * as an OSGi service or exposed remotely by using the request-respond
+ * features of Eclipse Scanning {@link https://github.com/eclipse/scanning}
+ * 
+ * There is an ISPyB specific implementation of this service in the 
+ * project https://github.com/DiamondLightSource/gda-ispyb-api. The class
+ * uk.ac.diamond.ispyb.scanning.ExperimentCommunicationService is the
+ * main/only implementation for ISPyB
+ * 
+ * @author Matthew Gerring
+ *
+ * @see {@link http://confluence.diamond.ac.uk/display/I151/GDA-Database+Communications+Concept}
+ * @see {@link http://confluence.diamond.ac.uk/display/I151/GDA-Database+Communications+Specification}
+ * 
+ */
+public interface IExperimentDatabaseService extends Closeable, ISampleDescriptionService {
+	
+
+	/**
+	 * Open the connection. 
+	 * A connection must be opened before it may be used
+	 * A service may be open and closed multiple times
+	 * @throws SQLException
+	 */
+	public void open() throws SQLException;
+	
+	/**
+	 * Close the connection to the database
+	 * A connection must be closed after use
+	 * A service may be open and closed multiple times
+	 * @throws IOException
+	 */
+	@Override
+	public void close() throws IOException;
+
+    /**
+     * This action is used to insert a record. Currently only a BeamlineAction may be inserted:
+	 <code><pre>
+		Bean beamlineAction = new Bean("uk.ac.diamond.ispyb.api.BeamlineAction");
+		beamlineAction.set("ProposalCode", "cm");
+		beamlineAction.set("ProposalNumber", 14451L);
+		beamlineAction.set("SessionNumber", 55167L);
+		beamlineAction.set("StartTime", new Timestamp(System.currentTimeMillis()));
+		beamlineAction.set("EndTime", new Timestamp(System.currentTimeMillis()));
+		beamlineAction.set("Message", "message");
+		beamlineAction.set("Parameter", "parameter");
+		beamlineAction.set("Value", "value");
+		beamlineAction.set("LogLevel", "DEBUG");
+		beamlineAction.set("Status", "PAUSED");
+		long code = service.insert(beamlineAction, true);
+	 </pre></code>
+     * 
+     * 
+     * @param action
+     * @param blocking
+ 	 * @return insert code or -1 if the upsert was asynchronous (non-blocking)
+     * @throws IllegalArgumentException
+     * @throws Exception
+     */
+    <T> Future<Id> insert(T action, boolean blocking)  throws IllegalArgumentException, Exception;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'java.lang.IllegalArgumentException' which is a subclass of 'java.lang.Exception'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)